packages/cli/src/lib/decorators/preconditions/preconditions.spec.ts

```diff
@@ -0,0 +1,92 @@
+import {Preconditions} from './preconditions';
+import type Command from '@oclif/command';
+
+describe('preconditions', () => {
+  const preconditions = new Array<jest.Mock<Boolean>>(5);
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('if all preconditions succeed', () => {
+    it('should executes preconditions the original command', async () => {
+      let counter = 0;
+      const orderChecker = jest.fn();
+      preconditions.fill(
+        jest.fn(
+          (() => {
+            orderChecker(counter++);
+            return true;
+          }).bind(this)
+        )
+      );
+
+      const fakeOriginalFunction = jest.fn(async () => {
+        orderChecker('final');
+      });
+      const fakeCommand = {};
+      const fakeDescriptor = {
+        value: fakeOriginalFunction,
+      };
+
+      await Preconditions(...preconditions)(
+        (fakeCommand as unknown) as Command,
+        '',
+        fakeDescriptor
+      );
+      await fakeDescriptor.value();
+
+      expect(orderChecker).toHaveBeenCalledTimes(preconditions.length + 1);
+      for (let index = 0; index < preconditions.length; index++) {
+        expect(orderChecker).toHaveBeenNthCalledWith(index + 1, index);
+      }
+      expect(fakeOriginalFunction).toHaveBeenCalled();
+      expect(orderChecker).toHaveBeenNthCalledWith(
+        preconditions.length + 1,
+        'final'
+      );
```

Comment:
I've heard somewhere that we should aim for one expect per test maximum.
I'm kidding. I understand the difference here :P
👌